indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/TombstoneHelper.java

```diff
@@ -173,66 +190,69 @@ public Set<DataSegment> computeTombstoneSegmentsForReplace(
    *                           They should be aligned with the replaceGranularity
    * @param dataSource         Datasource on which the replace is to be performed
    * @param replaceGranularity Granularity of the replace query
+   * @param maxBuckets         Maximum number of partition buckets. If the number of computed tombstone buckets
+   *                           exceeds this threshold, the method will throw an error.
    * @return Intervals computed for the tombstones
    * @throws IOException
    */
   public Set<Interval> computeTombstoneIntervalsForReplace(
       List<Interval> intervalsToDrop,
       List<Interval> intervalsToReplace,
       String dataSource,
-      Granularity replaceGranularity
+      Granularity replaceGranularity,
+      int maxBuckets
   ) throws IOException
   {
     Set<Interval> retVal = new HashSet<>();
     List<Interval> usedIntervals = getExistingNonEmptyIntervalsOfDatasource(intervalsToReplace, dataSource);
+    int buckets = 0;
 
     for (Interval intervalToDrop : intervalsToDrop) {
       for (Interval usedInterval : usedIntervals) {
 
         Interval overlap = intervalToDrop.overlap(usedInterval);
 
-        // No overlap of the dropped segment with the used interval due to which we donot need to generate any tombstone
+        // No overlap of the dropped segment with the used interval due to which we do not need to generate any tombstone
         if (overlap == null) {
           continue;
         }
+        if (buckets > maxBuckets) {
+          throw new IAE("Cannot add more tombstone buckets than [%d].", maxBuckets);
```

Comment:
I considered just returning the `TooManyBucketsException` directly here. But since this utility resides is in the common indexing package, we cannot use an MSQ fault `TooManyBucketsException`  here. So it has to be a more generic exception or introduce other semantics.